### PR TITLE
ERR: Clarify exceptions for invalid datetimelike operations

### LINKED ISSUE #9631: ERR: clarify exceptions for invalid datetimelike ops
xref #9630

IIRC most of these are `TypeErrors` (see `tseries/tests_base.py`, section on addition/subtraction ops)
but I guess *some* are`NotImplementedError``

```
In [3]: pd.date_range('20130101',periods=5)+Timestamp('20140101')
NotImplementedError: 
```

```
In [4]: pd.date_range('20130101',periods=5)+Timestamp('20140101').to_datetime64()

TypeError: ufunc add cannot use operands with types dtype('<M8[ns]') and dtype('<M8[ns]')
```

### PR DESCRIPTION
**Pull request by mturzanska**
closes #9631

### CODE REVIEW

**Comment by mturzanska** [comment]
Seemed too small a change for a what's new note

File: pandas/tests/indexes/datetimes/test_ops.py [hunk 1]
```diff
@@ -434,30 +434,53 @@ def test_add_iadd(self):
...
-        msg = "cannot add a datelike to a DatetimeIndex"
+        msg = "cannot add DatetimeIndex and Timestamp"
         with tm.assert_raises_regex(TypeError, msg):
             idx + Timestamp('2011-01-01')
 
         with tm.assert_raises_regex(TypeError, msg):
             Timestamp('2011-01-01') + idx
 
+    def test_add_dti_ts(self):
+        dti = DatetimeIndex(['2011-01-01', '2011-01-02'])
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_ops.py]: add the issue number as a comment

File: pandas/tests/indexes/datetimes/test_ops.py [hunk 2]
```diff
@@ -434,30 +434,53 @@ def test_add_iadd(self):
...
+        ts = Timestamp('2011-01-01')
+        msg = 'cannot add DatetimeIndex and Timestamp'
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_ops.py]: can you combine these into a single tests and use parametrize for the args

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
some comments. ping when pushed and green. skipping whatsnew is fine.

**Comment by mturzanska** [comment]
@jreback addressed all comments, tests pass

File: pandas/tests/indexes/datetimes/test_ops.py [hunk 3]
```diff
@@ -434,31 +434,29 @@ def test_add_iadd(self):
...
-
-        with pytest.raises(TypeError):
-            dti + dti
-
-        with pytest.raises(TypeError):
-            dti_tz + dti_tz
-
-        with pytest.raises(TypeError):
-            dti_tz + dti
-
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_ops.py]: hmm I think you removed the ``dti_tz + dti_tz`` tests (IOW a tz-aware interacting with a non-tzaware). these can be a separate test function. also add a datetime scalar addend.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/indexes/datetimes/test_ops.py [hunk 4]
```diff
@@ -458,6 +459,20 @@ def test_add_datetimelike(self, addend):
...
+    def test_add_dti_dti(self):
+        # previously performed setop (deprecated in 0.16.0)
+        # now raises TypeError (GH14164)
+
+        dti_tz = date_range('20110101', periods=3) \
+            .tz_localize('US/Eastern')
+        dti = date_range('20110101', periods=3)
+        combinations = product((dti_tz, dti), repeat=2)
+
+        msg = 'cannot add DatetimeIndex and DatetimeIndex'
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_ops.py]: can u parametrize this as well

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @mturzanska

**Pull request merged**